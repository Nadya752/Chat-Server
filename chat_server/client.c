#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <dirent.h>
#include <unistd.h>
#include <string.h>
#include <sys/stat.h>
#include <dirent.h>
#include "client.h"

// This file is for testing purposes only.

// Creates messages of client when they wants to SAYCONT in testing.
void saycont(char* id_name, char* domain_name, char* newmsg, int num){

    // Testing client will say "Hello I'm <id>", 
    // "Nice to meet you all", "I'm from <domain>".
    unsigned char type = SAYCONT;
    newmsg[0] = type;
    newmsg[1] =  0;

    if (num == 0){
        strcpy((char*) newmsg+START_PKT, "Hello I'm ");
        strcat((char*) newmsg+MSG_LEN1, id_name);
    }else if (num == 1){
        strcpy((char*) newmsg+START_PKT, "Nice to meet you all");
    }else {
        strcpy((char*) newmsg+START_PKT, "I'm from ");
        strcat((char*) newmsg+MSG_LEN2, domain_name);
        newmsg[MAX_BUF-1] = (char ) TERM;
    }

}

// Creates messages of client when they wants to SAY in testing.
void say(char* id_name, char* newmsg){

    // Testing client will say "Hello I'm <id>".
    unsigned char type = SAY;
    newmsg[0] = type;
    newmsg[1] =  0;
    strcpy(newmsg+START_PKT, "Hello I'm ");
    strcat(newmsg+MSG_LEN1, id_name);
}

// Creates messages of client when they wants to CONNECT.
void connect(char* id_name, char* domain_name, char* newmsg){
   unsigned char type = CONNECT;

    newmsg[0] = type;
    newmsg[1] =  0;
    short nidx = START_PKT;
    short ididx = 0;
    short didx = 0;
    while (id_name[ididx] != '\0' && ididx<= MAX_ID){
        newmsg[nidx] = id_name[ididx];
        nidx++;
        ididx++;

    }

    nidx = END_ID;
    while (domain_name[didx] != '\0' && didx <= (MAX_BUF - MAX_ID - START_PKT)){
        newmsg[nidx] = domain_name[didx];
        nidx++;
        didx++;

    }

}

// Checks if gevent pipe exists.
int check_gevent(){
    DIR* dir;
    struct dirent* folder;
    char filename[MAX_PNAME];
    char pipe_name[] = "gevent";
    dir = opendir(".");

    if (dir){
        while((folder= readdir(dir)) != NULL){
            memset(filename, 0,MAX_PNAME* sizeof(filename[0]));
            strcpy(filename, folder->d_name);
            if (strcmp(filename, pipe_name) == 0){
                printf("Pipe gevent is open\n");
                return TRUE;
            }
        }

    }
    printf("Pipe gevent is missing\n");
    return FALSE;
}

// Checks if a given domain exists.
int check_domain(char* dom){
    DIR* dir;
    dir = opendir(dom);

    if (dir){
        closedir(dir);
        printf("Domain %s exists\n", dom);
        return TRUE;
    }
    printf("Domain %s does not exists\n", dom);
    return FALSE;
}

// Checks if _RD and _WR pipes of given id in domain exists.
int check_rd_wr_pipe(char* dom, char* pipe_rd, char* pipe_wr){
    int has_pread = FALSE;
    int has_pwrite = FALSE;
    DIR* dir;
    struct dirent* folder;
    char filename[MAX_PNAME];
    dir = opendir(dom);

    // Open domain folder and search pipes.
    if (dir){
        while((folder= readdir(dir)) != NULL){
            memset(filename, 0,MAX_PNAME* sizeof(filename[0]));
            strcpy(filename, folder->d_name);
            if (strcmp(filename, pipe_rd) == 0){
                printf("Pipe %s is open\n", pipe_rd);
                has_pread = TRUE;
            }else if (strcmp(filename, pipe_wr)== 0){
                printf("Pipe %s is open\n", pipe_wr);
                has_pwrite = TRUE;
            }
        }

    }else{
        printf("failed opening domain\n");
    }
    
    return (has_pread && has_pwrite);

}
