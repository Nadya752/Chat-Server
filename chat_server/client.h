#ifndef CLIENT_H
#define CLIENT_H

#define MAX_ID (256)
#define TERM (255)
#define MAX_BUF (2048)
#define END_ID (258)
#define START_PKT (2)
#define MAX_PNAME (262)
#define MSG_LEN1 (10)
#define MSG_LEN2 (11)

typedef enum{
    CONNECT,
    SAY,
    SAYCONT,
    RECEIVE,
    RECVCONT,
    PING,
    PONG,
    DISCONNECT

}protocol;

typedef enum{
    FALSE,
    TRUE
    
} boolean;

int check_gevent();
int check_domain(char* dom);
int check_rd_wr_pipe(char* dom, char* pipe_rd, char* pipe_wr);
void connect(char* id_name, char* domain_name, char* newmsg);
void say(char* id_name, char* newmsg);
void saycont(char* id_name, char* domain_name, char* newmsg, int num);

#endif
