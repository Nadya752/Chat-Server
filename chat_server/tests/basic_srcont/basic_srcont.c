#include "../../client.h"
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <dirent.h>
#include <unistd.h>
#include <string.h>
#include <sys/stat.h>
#include <dirent.h>
#include <poll.h>

// Checks when a client SAYCONTs a message.
int main(){

    // Client named Ron connects.
    char conmsg1[MAX_BUF] = {0};
    connect("Ron", "Gryffindor", conmsg1);
    int fdcon1 = open("gevent", O_WRONLY);
    if (fdcon1 == -1){
        printf("Failed opening gevent.\n");
    }
    write(fdcon1, conmsg1, MAX_BUF);
    close(fdcon1);

    // Client named Harry connects.
    char conmsg2[MAX_BUF] = {0};
    connect("Harry", "Gryffindor", conmsg2);
    int fdcon2 = open("gevent", O_WRONLY);
    if (fdcon2 == -1){
        printf("Failed opening gevent.\n");
    }
    write(fdcon2, conmsg2, MAX_BUF);

    sleep(1);
    // Creating SAYCONT messages for Ron.
    char contmsg1[MAX_BUF] = {0};
    char contmsg2[MAX_BUF] = {0};
    char contmsg3[MAX_BUF] = {0};
    saycont("Ron", "Gryffindor", contmsg1, 0);
    saycont("Ron", "Gryffindor", contmsg2, 1);
    saycont("Ron", "Gryffindor", contmsg3, 2);

    // Ron sends SAYCONTs messages.
    ssize_t pwrite = open("./Gryffindor/Ron_WR", O_WRONLY);
    int wrres = write(pwrite,contmsg1, MAX_BUF);
    if(wrres == -1){
        fprintf(stderr, "%s", "Can't write pipe ");
    }
    if (pwrite == -1){
        perror("open()");
        fprintf(stderr, "%s", "Can't open write pipe\n");
    }
    close(pwrite);
    
    ssize_t pwrite2 = open("./Gryffindor/Ron_WR", O_WRONLY);
    int wrres2 = write(pwrite2,contmsg2, MAX_BUF);
    if(wrres2 == -1){
        fprintf(stderr, "%s", "Can't write pipe ");
    }
    if (pwrite2 == -1){
        perror("open()");
        fprintf(stderr, "%s", "Can't open write pipe\n");
    }
    close(pwrite2);

    ssize_t pwrite3 = open("./Gryffindor/Ron_WR", O_WRONLY);
    int wrres3 = write(pwrite3,contmsg3, MAX_BUF);
    if(wrres3 == -1){
        fprintf(stderr, "%s", "Can't write pipe ");
    }
    if (pwrite3 == -1){
        perror("open()");
        fprintf(stderr, "%s", "Can't open write pipe\n");
    }
    close(pwrite3);

    sleep(3);
    // Checks if Harry recieves all the SAYCONTs messages, including the termination byte.
    char newmsg[MAX_BUF] = {0};
    for (int i = 0; i<3; i++){
        ssize_t pread = open("./Gryffindor/Harry_RD", O_RDONLY);
        read(pread,newmsg, MAX_BUF);
        for(int idx = START_PKT; idx < MAX_BUF; idx++){
            if (newmsg[idx] != 0){
                if (idx == MAX_BUF-1){
                    printf("%x", newmsg[idx]);
                }else{
                    printf("%c", newmsg[idx]);
                }

            }

        }
        printf("\n");
        memset(newmsg, 0,MAX_BUF* sizeof(newmsg[0]));
        close(pread);
    }

    return 0;
    
}