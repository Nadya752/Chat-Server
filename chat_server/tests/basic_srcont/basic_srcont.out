RonHello I'm Ron
RonNice to meet you all
RonI'm from Gryffindorffffffff
